library/ssl_tls.c

```diff
@@ -1850,15 +1848,16 @@ int mbedtls_ssl_psk_derive_premaster( mbedtls_ssl_context *ssl, mbedtls_key_exch
 {
     unsigned char *p = ssl->handshake->premaster;
     unsigned char *end = p + sizeof( ssl->handshake->premaster );
-    const unsigned char *psk = ssl->conf->psk;
-    size_t psk_len = ssl->conf->psk_len;
+    const unsigned char *psk;
+    size_t psk_len;
 
-    /* If the psk callback was called, use its result */
-    if( ssl->handshake->psk != NULL )
-    {
-        psk = ssl->handshake->psk;
-        psk_len = ssl->handshake->psk_len;
-    }
+    if( mbedtls_ssl_get_psk( ssl, &psk, &psk_len )
+            == MBEDTLS_ERR_SSL_PRIVATE_KEY_REQUIRED )
+        /*
+         * This should never happen because the existence of a PSK is always
+         * checked before calling this function
+         */
+        MBEDTLS_SSL_DEBUG_MSG(1, ("should never happen"));
```

Comment:
What about to return in error?

Reply:
Damn, that's an excellent point! At the end of the function we're doing `memcpy( p, psk, psk_len );` and we definitely don't want to reach that line with `p == NULL`. Also, "should never happen" messages should always be followed immediately by `return( MBEDTLS_ERR_SSL_INTERNAL_ERROR )` so I'm not sure how I missed that.

Reply:
Fixed. Thanks for spotting!